Sources/XCTest/Public/XCTestCase.swift

```diff
@@ -108,41 +110,130 @@ open class XCTestCase: XCTest {
+        Task {
+            defer { asyncSetUpExpectation.fulfill() }
+            do {
+                try await setUp()
+            } catch {
+                recordError(error)
+                self.shouldInvokeTestClosure = false
             }
+        }
+        _ = XCTWaiter.wait(for: [asyncSetUpExpectation], timeout: Self.asyncTestTimeout)
 
-            if error.xct_shouldRecordAsTestSkip {
-                if let skip = error as? XCTSkip {
-                    self.skip = skip
-                } else {
-                    self.skip = XCTSkip(error: error, message: nil, sourceLocation: nil)
+        do {
+            try setUpWithError()
+        } catch {
+            recordError(error)
+            self.shouldInvokeTestClosure = false
+        }
+
+        setUp()
+    }
+    
+    func performTest() {
+        if shouldInvokeTestClosure {
+            switch testClosure {
+            case .sync(let closure):
+                do {
+                    try closure(self)
+                } catch {
+                    recordError(error)
                 }
+            case .async(let closure):
+                let testClosureExpectation = XCTestExpectation(description: "testClosureExpectation")
+                Task {
+                    defer { testClosureExpectation.fulfill() }
+                    do {
+                        try await closure(self)
+                    } catch {
+                        recordError(error)
+                    }
+                }
+                _ = XCTWaiter.wait(for: [testClosureExpectation], timeout: Self.asyncTestTimeout)
             }
         }
+    }
+    
+    func performTeardown() {
+        @Sendable func runTeardownBlocks(errorHandler: (Error) -> Void) async {
+            for block in await teardownBlocks.finalize().reversed() {
+                do {
+                    try await block()
+                } catch {
+                    errorHandler(error)
+                }
+            }
+        }
+        
+        let runTeardownBlocksExpectation = XCTestExpectation(description: "runTeardownBlocksExpectation")
+        Task {
+            defer { runTeardownBlocksExpectation.fulfill() }
+            await runTeardownBlocks(errorHandler: recordError(_:))
+        }
+        _ = XCTWaiter.wait(for: [runTeardownBlocksExpectation], timeout: Self.asyncTestTimeout)
 
-        if let skip = skip {
-            testRun?.recordSkip(description: skip.summary, sourceLocation: skip.sourceLocation)
+        // tearDown sequence
+        tearDown()
+
+        do {
+            try tearDownWithError()
+        } catch {
+            recordError(error)
         }
 
-        performTearDownSequence()
+        let asyncTeardownExpectation = XCTestExpectation(description: "asyncTeardownExpectation")
+        Task {
+            defer { asyncTeardownExpectation.fulfill() }
+            try await tearDown()
```

Comment:
We'll need to wrap this in a `do/catch` and call `recordError` like the other, non-async tearDown API calls above